plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java

```diff
@@ -245,48 +244,52 @@ public ListenableFuture<Void> createEmptyFile(URI file)
-                StorageBatch batch = storage.batch();
-                for (URI dir : directories) {
-                    Page<Blob> blobs = storage.list(getBucketName(dir), Storage.BlobListOption.prefix(keyFromUri(dir)));
-                    for (Blob blob : blobs.iterateAll()) {
-                        batch.delete(blob.getBlobId());
-                    }
-                }
-                batch.submit();
-            })));
+            return deleteReursivelyGcp(directories);
+        }
+
+        ImmutableMultimap.Builder<String, ListenableFuture<List<String>>> bucketToListObjectsFuturesBuilder = ImmutableMultimap.builder();
+        for (URI dir : directories) {
+            ImmutableList.Builder<String> keys = ImmutableList.builder();
+            ListenableFuture<List<String>> listObjectsFuture = Futures.transform(
+                    toListenableFuture((listObjectsRecursively(dir)
+                            .subscribe(listObjectsV2Response -> listObjectsV2Response.contents().stream()
+                                    .map(S3Object::key)
+                                    .forEach(keys::add)))),
+                    ignored -> keys.build(),
+                    directExecutor());
+            bucketToListObjectsFuturesBuilder.put(getBucketName(dir), listObjectsFuture);
+        }
+        Multimap<String, ListenableFuture<List<String>>> bucketToListObjectsFutures = bucketToListObjectsFuturesBuilder.build();
+
+        ImmutableList.Builder<ListenableFuture<List<DeleteObjectsResponse>>> deleteObjectsFutures = ImmutableList.builder();
+        for (String bucketName : bucketToListObjectsFutures.keySet()) {
+            deleteObjectsFutures.add(Futures.transformAsync(
+                    Futures.allAsList(bucketToListObjectsFutures.get(bucketName)),
+                    keys -> deleteObjects(
+                            bucketName,
+                            keys.stream()
+                                    .flatMap(Collection::stream)
+                                    .collect(toImmutableList())),
+                    directExecutor()));
         }
-        else {
-            ImmutableMultimap.Builder<String, ListenableFuture<List<String>>> bucketToListObjectsFuturesBuilder = ImmutableMultimap.builder();
+        return translateFailures(Futures.allAsList(deleteObjectsFutures.build()));
+    }
+
+    private ListenableFuture<Void> deleteReursivelyGcp(List<URI> directories)
+    {
+        // GCS is not compatible with S3's multi-object delete API https://cloud.google.com/storage/docs/migrating#methods-comparison
+        Storage storage = gcsClient.orElseThrow(() -> new IllegalStateException("gcsClient is expected to be initialized"));
+        ListeningExecutorService deleteExecutor = gcsDeleteExecutor.orElseThrow(() -> new IllegalStateException("gcsDeleteExecutor is expected to be initialized"));
+        return stats.getDeleteRecursively().record(asVoid(deleteExecutor.submit(() -> {
```

Comment:
I forgot this one, help me correct here
```suggestion
        return stats.getDeleteRecursively().record(translateFailures(deleteExecutor.submit(() -> {
```